src/compiler.jl

```diff
@@ -9448,33 +9514,47 @@ function _link(job, (mod, adjoint_name, primal_name, ctx, TapeType))
 end
 
 # actual compilation
-function _thunk(job, ctx=nothing, postopt=true)
-    # TODO: on 1.9, this actually creates a context. cache those.
-    if ctx === nothing
-        ctx = JuliaContext()
-    end
-    mod, meta = codegen(:llvm, job; optimize=false, ctx)
+function _thunk(job, postopt::Bool=true)
 
-    adjointf, augmented_primalf = meta.adjointf, meta.augmented_primalf
+    function __thunk()
+        mod, meta = codegen(:llvm, job; optimize=false)
 
-    adjoint_name = name(adjointf)
+        adjointf, augmented_primalf = meta.adjointf, meta.augmented_primalf
 
-    if augmented_primalf !== nothing
-        primal_name = name(augmented_primalf)
-    else
-        primal_name = nothing
+        adjoint_name = name(adjointf)
+
+        if augmented_primalf !== nothing
+            primal_name = name(augmented_primalf)
+        else
+            primal_name = nothing
+        end
+
+        LLVM.ModulePassManager() do pm
+            add!(pm, FunctionPass("ReinsertGCMarker", reinsert_gcmarker_pass!))
+            run!(pm, mod)
+        end
+        
+        # Run post optimization pipeline
+        if postopt && job.config.params.ABI <: FFIABI
+            post_optimze!(mod, JIT.get_tm())
+        end
+        return (mod, adjoint_name, primal_name, meta.TapeType)
     end
- 
-    LLVM.ModulePassManager() do pm
-        add!(pm, FunctionPass("ReinsertGCMarker", reinsert_gcmarker_pass!))
-        run!(pm, mod)
+
+@static if VERSION < v"1.9.0"
+    JuliaContext() do ctx
```

Comment:
Why don't we just make the context at the code that calls _thunk without it (e.g. from the generators)?

Reply:
I tried several permutations of this and always break with a missing ThreadSafeContext() in orcv2/get_trampoline.